### PR TITLE
[horovod] adding horovod init action

### LINKED ISSUE #775: Horovod initialization script
Hi,
I am trying to install Horovod on Dataproc nodes. Here are the steps I am following in my initialization actions script. However, I am not able to get openmpi or horovod installed on the worker nodes. It seems like the error is with the openmpi install steps. Manually doing these steps by ssh into worker nodes works.

To install openmpi (following instructions from [here](https://www.open-mpi.org/faq/?category=building#easy-build))
```
wget https://download.open-mpi.org/release/open-mpi/v4.0/openmpi-4.0.4.tar.gz
 tar -xzf openmpi-4.0.4.tar.gz
 cd /usr/lib/openmpi-4.0.4
 sudo ./configure --prefix=/usr/local
 sudo make all install
 export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/lib
```
tp install Horovod (following instructions from [here](https://github.com/horovod/horovod#install))
```
HOROVOD_GPU_OPERATIONS=NCCL pip install --user horovod
```

Could any instructions/pointers be provided to install Horovod on Dataproc nodes (with GPU support)?
Appreciate any help!

Thanks,

### PR DESCRIPTION
**Pull request by bradmiro**
Initialization to support Horovod on Dataproc.

Fixes #775 #779

### CODE REVIEW

**Comment by bradmiro** [comment]
Building Horovod with MPI leads to the following Dataproc initialization failure:

```
r2 failed to activate}'
<13>Jan 26 18:25:47 google-dataproc-startup[2907]: <13>Jan 26 18:25:47 activate-component-hive-server2[4323]: StructuredError{hive-server2, Component hive-server2 failed to activate}
<13>Jan 26 18:25:47 google-dataproc-startup[2907]: <13>Jan 26 18:25:47 activate-component-hive-server2[4323]: + exit 1
<13>Jan 26 18:25:47 google-dataproc-startup[2907]: + status=1
<13>Jan 26 18:25:47 google-dataproc-startup[2907]: + (( status != 0 ))
<13>Jan 26 18:25:47 google-dataproc-startup[2907]: + echo 'ERROR: Command cmd=[activate_component hive-server2] pid=4323 exited with 1'
<13>Jan 26 18:25:47 google-dataproc-startup[2907]: + tee /tmp/dataproc/commands/4323.error
<13>Jan 26 18:25:47 google-dataproc-startup[2907]: ERROR: Command cmd=[activate_component hive-server2] pid=4323 exited with 1
<13>Jan 26 18:25:47 google-dataproc-startup[2907]: + exit 1
```

File: horovod/BUILD [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        "scripts/verify_pytorch.py",
+        "scripts/verify_tensorflow.py",
+    ],
+    local = True,
+    shard_count = 4,
+    deps = [
+        "//integration_tests:dataproc_test_case",
+        "@io_abseil_py//absl/testing:parameterized",
+    ],
+)
```
Comment (medb) [inline horovod/BUILD:22]: And a new line at the end.

File: horovod/README.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
+# Dataproc - Horovod
```
Comment (medb) [inline horovod/README.md:1]: I think we can remove `Dataproc` from the title

File: horovod/README.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+take advantage of a series of Estimators provided by Horovod. Horovod is able
+to run on top of Spark, allowing you to scale easily with features such as
+Dataproc autoscaling.
+
+This initialization action will set up Horovod on a Google Cloud Dataproc
+cluster. You may configure Horovod to distribute its workloads using either
+MPI, Gloo, or NCCL with GPUS.
+
+Using this initialization action
+⚠️ NOTICE: See best practices of using initialization actions in production.
```
Comment (medb) [inline horovod/README.md:14]: I think that other init actions link to some doc

File: horovod/horovod.sh [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+readonly CUDA_VERSION
+
+HOROVOD_ENV_VARS="$(/usr/share/google/get_metadata_value attributes/horovod-env-vars || echo "")"
+readonly HOROVOD_ENV_VARS
+
+INSTALL_MPI="$(/usr/share/google/get_metadata_value attributes/install-mpi || echo "")"
+readonly INSTALL_MPI
+
+function install_mpi() {
+  local -r MPI_VERSION="4.1.0"
```
Comment (medb) [inline horovod/horovod.sh:51]: Make local variables lower case

File: horovod/horovod.sh [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+  local -r MPI_URL="https://download.open-mpi.org/release/open-mpi/v4.1/openmpi-${MPI_VERSION}.tar.gz"
+  #apt install openmpi
```
Comment (medb) [inline horovod/horovod.sh:53]: Remove?

File: horovod/horovod.sh [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+  
+  tmp_dir=$(mktemp -d -t mlvm-horovod-mpi-XXXX)
```
Comment (medb) [inline horovod/horovod.sh:55]: Declare as local

File: horovod/horovod.sh [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+export CLASSPATH=$(hadoop classpath --glob)
+EOF
+}
+
+function install_frameworks() {
+  local frameworks=()
+
+  # Add gpu-versions of libraries
+  if (lspci | grep -q NVIDIA); then
+    pip install "torch==${PYTORCH_VERSION}+cu${CUDA_VERSION//.}" "torchvision==${TORCHVISION_VERSION}+cu${CUDA_VERSION//.}" -f "https://download.pytorch.org/whl/torch_stable.html"
```
Comment (medb) [inline horovod/horovod.sh:80]: Are all of these available in conda?

File: horovod/horovod.sh [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+      "mxnet==${MXNET_VERSION}"
+      )
+  fi
+
+  pip install "${frameworks[@]}"
+}
+
+function install_horovod() {
+  # Install cmake
+  apt-get install -y cmake
```
Comment (medb) [inline horovod/horovod.sh:100]: Add retries

File: horovod/horovod.sh [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+  # Install Horovod
+  horovod_build_frameworks="tensorflow,pytorch,spark"
```
Comment (medb) [inline horovod/horovod.sh:103]: Declare as local

File: horovod/horovod.sh [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+  
+  # Horovod on Dataproc with MXNet only supported with CPUs
+  if ! (lspci | grep -q NVIDIA); then
+    horovod_build_frameworks+=",mxnet"
+  fi
+
+  eval "${HOROVOD_ENV_VARS} pip install --no-cache-dir horovod[${horovod_build_frameworks}]==${HOROVOD_VERSION}"
```
Comment (medb) [inline horovod/horovod.sh:110]: Use `conda` if possible

File: horovod/horovod.sh [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+  configure_hadoop_env
+
+  # Install Frameworks
+  install_frameworks
+
+  # Install Horovod
+  install_horovod
+}
+
+main
```
Comment (medb) [inline horovod/horovod.sh:129]: Add EOF line

File: horovod/scripts/verify_mxnet.py [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+from pyspark.sql import SparkSession
+
+spark = SparkSession.builder.getOrCreate()
+def fn(magic_number):
+    import horovod.torch as hvd
+    hvd.init()
+    print('Hello, rank = %d, local_rank = %d, size = %d, local_size = %d, magic_number = %d' % (hvd.rank(), hvd.local_rank(), hvd.size(), hvd.local_size(), magic_number))
+    return hvd.rank()
+
+horovod.spark.run(fn, args=(42,))
```
Comment (medb) [inline horovod/scripts/verify_mxnet.py:11]: Add EOF line

File: horovod/scripts/verify_pytorch.py [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+from pyspark.sql import SparkSession
+
+spark = SparkSession.builder.getOrCreate()
+def fn(magic_number):
+    import horovod.torch as hvd
+    hvd.init()
+    print('Hello, rank = %d, local_rank = %d, size = %d, local_size = %d, magic_number = %d' % (hvd.rank(), hvd.local_rank(), hvd.size(), hvd.local_size(), magic_number))
+    return hvd.rank()
+
+horovod.spark.run(fn, args=(42,))
```
Comment (medb) [inline horovod/scripts/verify_pytorch.py:11]: Add EOF line

File: horovod/scripts/verify_tensorflow.py [hunk 14]
```diff
@@ -0,0 +1,11 @@
...
+from pyspark.sql import SparkSession
+
+spark = SparkSession.builder.getOrCreate()
+def fn(magic_number):
+    import horovod.tensorflow as hvd
+    hvd.init()
+    print('Hello, rank = %d, local_rank = %d, size = %d, local_size = %d, magic_number = %d' % (hvd.rank(), hvd.local_rank(), hvd.size(), hvd.local_size(), magic_number))
+    return hvd.rank()
+
+horovod.spark.run(fn, args=(42,))
```
Comment (medb) [inline horovod/scripts/verify_tensorflow.py:11]: Add EOF line

**Review by medb — CHANGES_REQUESTED** [review]

File: horovod/horovod.sh (hunk 7 above)
Comment (bradmiro) [reply horovod/horovod.sh:80]: Pytorch is the only library here reliably available in Conda with the latest versions. MXNet, TensorFlow, and Horovod are not. I can try moving just PyTorch to Conda and see if that will work.

**Comment by bradmiro** [comment]
> Building Horovod with MPI leads to the following Dataproc startup failure:
> 
> ```
> <13>Jan 26 18:25:47 google-dataproc-startup[2907]: <13>Jan 26 18:25:47 activate-component-hive-server2[4323]: + [[ 1 -eq 2 ]]
> <13>Jan 26 18:25:47 google-dataproc-startup[2907]: <13>Jan 26 18:25:47 activate-component-hive-server2[4323]: + echo 'StructuredError{hive-server2, Component hive-server2 failed to activate}'
> r2 failed to activate}'
> <13>Jan 26 18:25:47 google-dataproc-startup[2907]: <13>Jan 26 18:25:47 activate-component-hive-server2[4323]: StructuredError{hive-server2, Component hive-server2 failed to activate}
> <13>Jan 26 18:25:47 google-dataproc-startup[2907]: <13>Jan 26 18:25:47 activate-component-hive-server2[4323]: + exit 1
> <13>Jan 26 18:25:47 google-dataproc-startup[2907]: + status=1
> <13>Jan 26 18:25:47 google-dataproc-startup[2907]: + (( status != 0 ))
> <13>Jan 26 18:25:47 google-dataproc-startup[2907]: + echo 'ERROR: Command cmd=[activate_component hive-server2] pid=4323 exited with 1'
> <13>Jan 26 18:25:47 google-dataproc-startup[2907]: + tee /tmp/dataproc/commands/4323.error
> <13>Jan 26 18:25:47 google-dataproc-startup[2907]: ERROR: Command cmd=[activate_component hive-server2] pid=4323 exited with 1
> <13>Jan 26 18:25:47 google-dataproc-startup[2907]: + exit 1
> ```
> 
> Code:
> https://github.com/GoogleCloudDataproc/initialization-actions/blob/69c5ec1aaaa87c9ce3e001bd2ff374deb93ba023/horovod/horovod.sh#L50-L65

I'm noticing a similar error with the MLVM. Could this be related to the fact that init actions are now run before the Dataproc startup script?

File: horovod/horovod.sh (hunk 10 above)
Comment (bradmiro) [reply horovod/horovod.sh:110]: There's no reliable release of Horovod on any Conda channel

**Review by medb — APPROVED** [review]

**Pull request merged**